infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServerNode.java

```diff
@@ -90,6 +89,7 @@ public CorfuServerNode(@Nonnull ServerContext serverContext,
         this.serverMap = serverMap;
         router = new NettyServerRouter(new ArrayList<>(serverMap.values()));
         this.serverContext.setServerRouter(router);
+        this.serverContext.setServerEpochForSingleNodeLayout(serverContext.getServerEpoch(), router);
```

Comment:
I dont think this is needed.

The root cause of the problem is: the epoch's uninitialized value is a valid value. Therefore, we need to change `ServerContext:getServerEpoch` to return an invalid value if epoch == null. Instead of returning 0 for null, we should return `Layout.INVALID_EPOCH`

Reply:
I tried doing just that but a lot of tests failed, since it's necessary for some servers to have the context server epoch to start at 0.

Reply:
It possible to pass -1 to the BatchProcessor on start up, which will probably work, but then the client would see a WrongEpochException(-1), which is misleading, since they write to the log unit with the epoch 0. So I thought the better approach would be to start with None Epoch and then just seal it after the bootstrap. If the node was never bootstrapped, it was never sealed, so all the writes are rejected.

Reply:
I still think the uninitialized epoch value shouldn't have a valid epoch number.

But even if that is done, I don't think its a generic way to deal with this issue. If you look at `LayoutServer` , you'll see that each handler (except bootstrap) will try to check `isBootstrapped`. Essentially, its the same problem: If a server is not bootstrapped it shouldn't service RPCs. I think the cleanest way to do this is at the message routing level. The `NettyServerRouter` should reject all messages as long as the server hasn't been bootstrapped.

Also, I don;t think its necessary to add a new exception type. `NoBootstrapException` is exactly what you need.

Reply:
We might also need to check clusterId on BaseServer.

Reply:
@Maithem sounds good. 
@WenbinZhu that sounds good, it will entail adding clusterId as one of the fields of CorfuMsg, some changes with respect to serialization/deserialization and checking of the clusterId from the router. Hopefully, not a lot of work